File: javascript/poker/poker.js
const RANKS = '23456789TJQKA';
const SUITS = 'SCDH';

// Parses a card string (e.g., 'KH', '10S') into rank and suit values
const parseCard = (cardStr) => {
  const suitStr = cardStr.slice(-1);
  const rankStr = cardStr.length === 3 && cardStr.startsWith('10') ? 'T' : cardStr.slice(0, -1); // Handle '10' specifically
  const rank = RANKS.indexOf(rankStr);
  const suit = SUITS.indexOf(suitStr);
  if (rank === -1 || suit === -1 || (cardStr.length === 3 && !cardStr.startsWith('10'))) { // Added check for invalid length like '100S'
    throw new Error(`Invalid card: ${cardStr}`);
  }
  // Return the original rank string ('10') if it was 10, otherwise the single char
  const originalRankStr = cardStr.length === 3 && cardStr.startsWith('10') ? '10' : rankStr;
  return { rank, suit, rankStr: originalRankStr, suitStr };
};

// Parses a hand string (e.g., '4S 5S 7H 8D JC') into sorted cards
const parseHand = (handStr) => {
  const cards = handStr.split(' ').map(parseCard);
  if (cards.length !== 5) {
    throw new Error(`Invalid hand size: ${handStr}`);
  }
  // Sort cards by rank descending
  cards.sort((a, b) => b.rank - a.rank);
  return cards;
};

// Gets counts of each rank and suit
const getCounts = (cards) => {
  const rankCounts = Array(RANKS.length).fill(0);
  const suitCounts = Array(SUITS.length).fill(0);
  const ranks = [];
  cards.forEach(card => {
    rankCounts[card.rank]++;
    suitCounts[card.suit]++;
    ranks.push(card.rank);
  });
  return { rankCounts, suitCounts, ranks };
};

// Determines the rank and tie-breaking values for a hand
const getHandDetails = (handStr) => {
  const cards = parseHand(handStr);
  const { rankCounts, suitCounts, ranks } = getCounts(cards);

  const isFlush = suitCounts.some(count => count >= 5);
  // Check for Ace-low straight (A, 2, 3, 4, 5)
  const isAceLowStraight = ranks.toString() === '12,3,2,1,0'; // A, 5, 4, 3, 2 sorted descending
  const ranksForStraightCheck = isAceLowStraight ? [3, 2, 1, 0, -1] : ranks; // Use 5,4,3,2,A (represented by indices 3,2,1,0,-1) for Ace-low
  
  let isStraight = true;
  for (let i = 0; i < ranksForStraightCheck.length - 1; i++) {
    if (ranksForStraightCheck[i] !== ranksForStraightCheck[i+1] + 1) {
      isStraight = false;
      break;
    }
  }
  
  const straightHighCard = isAceLowStraight ? 3 : ranksForStraightCheck[0]; // High card is 5 (index 3) for A-low

  if (isStraight && isFlush) {
    return { rank: 8, values: [straightHighCard], handStr }; // Straight Flush (or Royal Flush if high card is Ace)
  }

  const counts = rankCounts.filter(count => count > 0).sort((a, b) => b - a);
  const rankGroups = rankCounts
    .map((count, rank) => ({ rank, count }))
    .filter(group => group.count > 0)
    .sort((a, b) => b.count - a.count || b.rank - a.rank); // Sort by count desc, then rank desc

  const groupValues = rankGroups.map(g => g.rank);

  if (counts[0] === 4) {
    return { rank: 7, values: groupValues, handStr }; // Four of a Kind
  }
  if (counts[0] === 3 && counts[1] === 2) {
    return { rank: 6, values: groupValues, handStr }; // Full House
  }
  if (isFlush) {
    return { rank: 5, values: ranks, handStr }; // Flush
  }
  if (isStraight) {
    return { rank: 4, values: [straightHighCard], handStr }; // Straight
  }
  if (counts[0] === 3) {
    return { rank: 3, values: groupValues, handStr }; // Three of a Kind
  }
  if (counts[0] === 2 && counts[1] === 2) {
    return { rank: 2, values: groupValues, handStr }; // Two Pair
  }
  if (counts[0] === 2) {
    return { rank: 1, values: groupValues, handStr }; // One Pair
  }
  return { rank: 0, values: ranks, handStr }; // High Card
};

// Compares two hands based on rank and tie-breaking values
const compareHands = (handA, handB) => {
  if (handA.rank !== handB.rank) {
    return handB.rank - handA.rank;
  }
  for (let i = 0; i < handA.values.length; i++) {
    if (handA.values[i] !== handB.values[i]) {
      return handB.values[i] - handA.values[i];
    }
  }
  return 0; // Hands are equal
};

export const bestHands = (hands) => {
  if (!hands || hands.length === 0) {
    return [];
  }

  const handDetails = hands.map(getHandDetails);
  handDetails.sort(compareHands);

  const bestRank = handDetails[0].rank;
  const bestValues = handDetails[0].values;

  const winningHands = handDetails
    .filter(hand => hand.rank === bestRank && compareHands(hand, handDetails[0]) === 0)
    .map(hand => hand.handStr);

  return winningHands;
};

File: go/poker/poker.go
package poker

import (
	"errors"
	"sort"
	"strings"
)

type card struct {
	rank int
	suit rune
}

type hand struct {
	cards []card
	raw   string
	rank  int
	tiebreakers []int
}

var rankMap = map[string]int{
	"2": 2, "3": 3, "4": 4, "5": 5, "6": 6, "7": 7, "8": 8, "9": 9, "10": 10,
	"J": 11, "Q": 12, "K": 13, "A": 14,
}

var validSuits = map[rune]bool{'♤': true, '♡': true, '♢': true, '♧': true}

func parseCard(s string) (card, error) {
	s = strings.TrimSpace(s)
	runes := []rune(s)
	if len(runes) < 2 {
		return card{}, errors.New("invalid card length")
	}
	suit := runes[len(runes)-1]
	rankStr := string(runes[:len(runes)-1])
	rankStr = strings.TrimSpace(rankStr)
	rank, ok := rankMap[rankStr]
	if !ok {
		return card{}, errors.New("invalid rank: " + rankStr)
	}
	if !validSuits[suit] {
		return card{}, errors.New("invalid suit")
	}
	return card{rank, suit}, nil
}

func parseHand(raw string) (hand, error) {
	parts := strings.Split(raw, " ")
	if len(parts) != 5 {
		return hand{}, errors.New("hand must have 5 cards")
	}
	cards := make([]card, 5)
	for i, p := range parts {
		c, err := parseCard(p)
		if err != nil {
			return hand{}, err
		}
		cards[i] = c
	}
	sort.Slice(cards, func(i, j int) bool { return cards[i].rank < cards[j].rank })
	h := hand{cards: cards, raw: raw}
	h.rank, h.tiebreakers = evaluateHand(cards)
	return h, nil
}

func evaluateHand(cs []card) (int, []int) {
	isFlush := true
	for i := 1; i < 5; i++ {
		if cs[i].suit != cs[0].suit {
			isFlush = false
			break
		}
	}
	ranks := make([]int, 5)
	for i, c := range cs {
		ranks[i] = c.rank
	}
	isStraight := true
	for i := 1; i < 5; i++ {
		if ranks[i] != ranks[i-1]+1 {
			isStraight = false
			break
		}
	}
	// Special case: Ace-low straight (A 2 3 4 5)
	if !isStraight && ranks[4] == 14 && ranks[0] == 2 && ranks[1] == 3 && ranks[2] == 4 && ranks[3] == 5 {
		isStraight = true
		ranks[4] = 1 // treat Ace as 1 for tie-breaking
		sort.Ints(ranks)
	}

	counts := make(map[int]int)
	for _, r := range ranks {
		counts[r]++
	}

	var pairs, trips, quads int
	var pairRanks, tripRanks, quadRanks []int
	for r, cnt := range counts {
		switch cnt {
		case 4:
			quads++
			quadRanks = append(quadRanks, r)
		case 3:
			trips++
			tripRanks = append(tripRanks, r)
		case 2:
			pairs++
			pairRanks = append(pairRanks, r)
		}
	}

	sort.Sort(sort.Reverse(sort.IntSlice(pairRanks)))
	sort.Sort(sort.Reverse(sort.IntSlice(tripRanks)))
	sort.Sort(sort.Reverse(sort.IntSlice(quadRanks)))

	switch {
	case isStraight && isFlush:
		return 8, []int{ranks[4]} // straight flush
	case quads == 1:
		kickers := getKickers(ranks, quadRanks)
		return 7, append(quadRanks, kickers...)
	case trips == 1 && pairs == 1:
		return 6, append(tripRanks, pairRanks...)
	case isFlush:
		return 5, reverseSort(ranks)
	case isStraight:
		return 4, []int{ranks[4]}
	case trips == 1:
		kickers := getKickers(ranks, tripRanks)
		return 3, append(tripRanks, kickers...)
	case pairs == 2:
		kickers := getKickers(ranks, pairRanks)
		return 2, append(pairRanks, kickers...)
	case pairs == 1:
		kickers := getKickers(ranks, pairRanks)
		return 1, append(pairRanks, kickers...)
	default:
		return 0, reverseSort(ranks)
	}
}

func getKickers(ranks []int, exclude []int) []int {
	m := make(map[int]bool)
	for _, e := range exclude {
		m[e] = true
	}
	var kickers []int
	for i := len(ranks) - 1; i >= 0; i-- {
		if !m[ranks[i]] {
			kickers = append(kickers, ranks[i])
		}
	}
	return kickers
}

func reverseSort(ranks []int) []int {
	out := make([]int, len(ranks))
	copy(out, ranks)
	sort.Sort(sort.Reverse(sort.IntSlice(out)))
	return out
}

func BestHand(hands []string) ([]string, error) {
	parsed := make([]hand, 0, len(hands))
	for _, raw := range hands {
		h, err := parseHand(raw)
		if err != nil {
			return nil, err
		}
		parsed = append(parsed, h)
	}
	bestRank := -1
	var bestHands []hand
	for _, h := range parsed {
		if h.rank > bestRank {
			bestRank = h.rank
			bestHands = []hand{h}
		} else if h.rank == bestRank {
			bestHands = append(bestHands, h)
		}
	}
	// tie-breaker
	var winners []string
	for _, h := range bestHands {
		if len(winners) == 0 {
			winners = []string{h.raw}
			continue
		}
		comp := compareTiebreakers(h.tiebreakers, parsedHandByRaw(winners[0], parsed).tiebreakers)
		if comp > 0 {
			winners = []string{h.raw}
		} else if comp == 0 {
			winners = append(winners, h.raw)
		}
	}
	return winners, nil
}

func parsedHandByRaw(raw string, hands []hand) hand {
	for _, h := range hands {
		if h.raw == raw {
			return h
		}
	}
	return hand{}
}

func compareTiebreakers(a, b []int) int {
	for i := 0; i < len(a) && i < len(b); i++ {
		if a[i] > b[i] {
			return 1
		} else if a[i] < b[i] {
			return -1
		}
	}
	return 0
}
